### PR TITLE
[PBI-1849] Add wrapper ad logic

### PR DESCRIPTION
**Pull request by dchauOoyala**

### CODE REVIEW

File: js/ad_manager_vast.js [hunk 1]
```diff
@@ -36,7 +36,7 @@ OO.Ads.manager(function(_, $) {
    * @property {string} adURLOverride If the page level params override the ad url then it is stored here
    * @property {object} lastOverlayAd Contains the ad information for the overlay that was displayed before it was removed.
    * This is used so we know what to add back to the screen after the video ad is done and the main video hasn't ended.
-   * @property {object} errorInfo The object that holds each individual ad id's error urls. Used for error reporting.
+   * @property {object} adTrackingInfo The object that holds each individual ad id's tracking urls.
```
Comment (pilievOoyala) [inline js/ad_manager_vast.js]: Can you add that this includes error tracking too?

File: js/ad_manager_vast.js [hunk 2]
```diff
@@ -1039,16 +1042,56 @@ OO.Ads.manager(function(_, $) {
...
+
+    /**
+     * Helper function to return the object that directly contains the VAST ad information.
+     * The name "adObject" or "ad" is a loaded name in our code. At different points of the
+     * plugin logic, these ad objects contain different properties/information (i.e. the "ad
+     * object" returned from AMC or the "ad object" returned from parsing the VAST ad response).
+     * The "ad object" returned from AMC has the VAST ad object nested in it.
+     * @private
+     * @method Vast#_getVastAdObject
+     * @param {object} adObject The object containing the VAST ad object, or the VAST ad object
```
Comment (pilievOoyala) [inline js/ad_manager_vast.js]: From our talk, do you want to rename this to amcAdObject for clarity?

File: js/ad_manager_vast.js [hunk 3]
```diff
@@ -1079,14 +1121,13 @@ OO.Ads.manager(function(_, $) {
...
+      adObject = _getVastAdObject(adObject);
       var linearClickTrackingUrls = null;
       if (adObject &&
-          adObject.ad &&
-          adObject.ad.data &&
-          adObject.ad.data.linear &&
-          adObject.ad.data.linear.clickTracking &&
-          adObject.ad.data.linear.clickTracking.length > 0) {
-        linearClickTrackingUrls = adObject.ad.data.linear.clickTracking;
+          adObject.linear &&
```
Comment (pilievOoyala) [inline js/ad_manager_vast.js]: nice work on shortening the if checks.

File: js/ad_manager_vast.js [hunk 4]
```diff
@@ -1195,41 +1232,77 @@ OO.Ads.manager(function(_, $) {
...
+
       return trackingUrls;
     };
 
+    /**
+     * Helper function to retrieve the ad object's ad ID (the value of the <Ad> element's "id" attribute).
+     * Not the "id" our ad manager assigns.
+     * @private
+     * @method Vast#_getAdId
+     * @param {object} adObject The ad metadata
```
Comment (pilievOoyala) [inline js/ad_manager_vast.js]: Is this amc ad object or the inner one?  If it's amc, can you rename it here too.

File: js/ad_manager_vast.js [hunk 5]
```diff
@@ -1195,41 +1232,77 @@ OO.Ads.manager(function(_, $) {
...
-    var _pingTrackingUrls = _.bind(function(urlObject) {
+    var _pingTrackingUrls = _.bind(function(urlObject, adId) {
       for (var trackingName in urlObject) {
         if (urlObject.hasOwnProperty(trackingName)) {
           try {
             var urls = urlObject[trackingName];
             if (urls) {
               OO.pixelPings(urls);
-              OO.log("VAST: \"" + trackingName + "\" tracking URLs pinged");
+              OO.log("VAST: \"" + trackingName + "\" tracking URLs pinged for VAST Ad Id: " + adId);
```
Comment (pilievOoyala) [inline js/ad_manager_vast.js]: I think this crashes if adId is undefined. Can you double check and add an if statement, if needed?

File: js/ad_manager_vast.js [hunk 6]
```diff
@@ -2324,6 +2401,28 @@ OO.Ads.manager(function(_, $) {
 
     }, this);
 
+    var _handleWrapperAd = _.bind(function(ad, adLoaded) {
```
Comment (pilievOoyala) [inline js/ad_manager_vast.js]: can you clarify if it's amc ad or not? and can you use _getAdID function you created?

File: test/unit-test-helpers/mock_responses/vast_wrapper_2.xml [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+  <Wrapper>
+    <AdTitle>wrapperParentAd2</AdTitle>
+    <VASTAdTagURI>vastad.xml</VASTAdTagURI>
+    <Impression>impressionUrl</Impression>
+    <Error>errorUrl</Error>
+	<Creatives>
+		<Creative AdID="wrapper-parent-creative-id-2">
+			<Linear>
+				<TrackingEvents>
+          <Tracking event="start">startUrl</Tracking>
```
Comment (pilievOoyala) [inline test/unit-test-helpers/mock_responses/vast_wrapper_2.xml]: would it be useful to call this startUrl2 or something, so you know you aren't firing the same events as in the first wrapper file?

File: js/ad_manager_vast.js (hunk 5 above)
Comment (dchauOoyala) [inline js/ad_manager_vast.js]: fortunately, if adId is undefined/null, it will just print "undefined"/"null"

File: test/unit-tests/vast_test.js [hunk 8]
```diff
@@ -2327,4 +2329,103 @@ describe('ad_manager_vast', function() {
...
+    var parentDepthOneObject = adTrackingInfo[parentDepthOneId];
+    var parentDepthTwoObject = adTrackingInfo[parentDepthTwoId];
+    var leafObject = adTrackingInfo[leafId];
+
+    // Ad Tracking Objects should have correct wrapper parent IDs
+    expect(parentDepthOneObject.wrapperParentId).to.be(null);
+    expect(parentDepthTwoObject.wrapperParentId).to.be(parentDepthOneId);
+    expect(leafObject.wrapperParentId).to.be(parentDepthTwoId);
+
+    expect(parentDepthOneObject.adObject).to.not.be(null);
```
Comment (pilievOoyala) [inline test/unit-tests/vast_test.js]: You are checking that the object exists. Can you check and see if some of the messages exist within the right object?

**Comment by pilievOoyala** [comment]
LGTM

**Pull request merged**